blockchain/blockdao.go

```diff
@@ -606,6 +652,82 @@ func (dao *blockDAO) deleteTipBlock() error {
 	return dao.kvstore.Commit(batch)
 }
 
+// getDBForHash returns db of this block stored
+func (dao *blockDAO) getDBFromHash(h hash.Hash256) (db.KVStore, int, error) {
+	blockHashDBKey := append(blockHashDBPrefix, h[:]...)
+	whichDBValue, err := dao.kvstore.Get(blockNS, blockHashDBKey)
+	if err != nil {
+		return dao.kvstore, 0, nil
+	}
+	index := int(enc.MachineEndian.Uint64(whichDBValue))
+
+	db, err := dao.getDBFromIndex(index)
+	if err != nil {
+		return nil, 0, err
+	}
+	return db, index, nil
+}
+
+// getDB get db if exists,or will create new db and return
+func (dao *blockDAO) getDBFromIndex(index int) (kvstore db.KVStore, err error) {
+	if index == 0 {
+		return dao.kvstore, nil
+	}
+	kv, ok := dao.kvstores.Load(index)
+	if ok {
+		kvstore, ok = kv.(db.KVStore)
+		if !ok {
+			err = errors.New("db convert error")
+		}
+		return
+	}
+
+	// create new db according to cfg's file name
+	var withSuffix, suffix, name string
+	cfg := dao.cfg
+	withSuffix = path.Base(cfg.DbPath)
+	suffix = path.Ext(withSuffix)
+	name = strings.TrimSuffix(withSuffix, suffix)
+	name += fmt.Sprintf("-%d", index) + ".db"
+	cfg.DbPath = path.Dir(cfg.DbPath) + "/" + name
+
+	kvstore = db.NewBoltDB(cfg)
+	dao.kvstores.Store(index, kvstore)
+	err = kvstore.Start(context.Background())
```

Comment:
A follow up may be closing the keystore when it's not been used for a while

Reply:
I  think it will close when call dao.lifecycle.Stop()